pkg/services/ngalert/state/manager.go

```diff
@@ -99,11 +110,54 @@ func (st *Manager) Run(ctx context.Context) error {
 		case <-ctx.Done():
 			st.log.Debug("Stopping")
 			ticker.Stop()
+			if st.saveStateAsync {
+				close(st.stateRunnerShutdown)
+				st.stateRunnerWG.Wait()
+			}
 			return ctx.Err()
 		}
 	}
 }
 
+func (st *Manager) startSync(ctx context.Context) {
+	st.stateRunnerWG.Add(1)
+	go func(ctx context.Context) {
+		ticker := st.clock.Ticker(time.Second * 30)
+	infLoop:
+		for {
+			select {
+			case <-ticker.C:
+				if err := st.fullSync(ctx); err != nil {
+					st.log.Error("Failed to do a full state sync to database", "err", err)
+				}
+			case <-st.stateRunnerShutdown:
+				st.log.Info("Stopping state sync...")
+				shutdownCtx, cancel := context.WithTimeout(context.Background(), time.Second*30)
```

Comment:
I'm not sure about how I feel about a hard shutdown of `30s` - Can we get any stats on shutdown times for Grafana pods? If not, I'd rather just leave this without a hard timeout. Otherwise, this is the equivalent of an unclean  shutdown and could potentially happen on every restart for bigger instances.